src/main/java/com/cloudbees/jenkins/support/util/Anonymizer.java

```diff
@@ -0,0 +1,207 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package com.cloudbees.jenkins.support.util;
+
+import com.cloudbees.jenkins.support.SupportPlugin;
+import hudson.Functions;
+import hudson.XmlFile;
+import hudson.model.Computer;
+import hudson.model.Item;
+import hudson.model.Label;
+import hudson.model.Node;
+import hudson.model.User;
+import hudson.model.View;
+import jenkins.model.Jenkins;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+import org.kohsuke.randname.RandomNameGenerator;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentSkipListMap;
+import java.util.concurrent.ConcurrentSkipListSet;
+import java.util.concurrent.CopyOnWriteArraySet;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+@Restricted(NoExternalUse.class)
+public class Anonymizer {
+    private static final Logger LOGGER = Logger.getLogger(Anonymizer.class.getName());
+    private static final RandomNameGenerator GENERATOR = new RandomNameGenerator();
+    private static final Set<String> SEPARATORS = new CopyOnWriteArraySet<>();
+    private static final Map<String, String> ANON_MAP = new ConcurrentHashMap<>();
+    private static final Set<String> ORDER = new ConcurrentSkipListSet<>(Comparator.comparingInt(String::length).reversed().thenComparing(s -> s));
+    private static final Map<String, String> DISPLAY = new ConcurrentSkipListMap<>();
+    private static final long DISPLAY_REFRESH_INTERVAL = 1000 * 60 * 10; // 10 mins
+
+    private static long LAST_REFRESH;
+
+    private Anonymizer() { }
+
+    static {
+        // Full names can have this separator
+        SEPARATORS.add(" » ");
+        refresh();
+    }
+
+    @SuppressWarnings("unchecked")
+    public static void refresh() {
+        LOGGER.log(Level.FINE, "Refreshing anonymized items");
+        Jenkins instance = Jenkins.getInstance();
+        load();
+
+        SupportPlugin.AnonymizationSettings settings = SupportPlugin.getInstance().getAnonymizationSettings();
+        if (settings.isAnonymizeLabels()) {
```

Comment:
We should probably refresh the labels after nodes, because otherwise all nodes end up with the `label` prefix since there is always an implicit label with the name of the agent.